File: src/io/rtt_manager.py
import re
import time
from src.io.io_transport import IOTransport
from src.io.rtt_reader import RttReaderThread


RTT_WRITE_TIMEOUT_SECONDS = 0.1
RTT_WRITE_RETRY_INTERVAL_SECONDS = 0.001


class RttManager(IOTransport):
    """RTT 管理类 - 封装 J-Link RTT 读写操作"""

    def __init__(self):
        super().__init__()
        self.jlink = None
        self.settings = {
            'chip': '',
            'speed': 4000,
            'reset': False,
            'start_address': '',
            'range_size': '',
            'frame_timeout': 50,
        }

    def _import_pylink(self, emit_error=True):
        try:
            import pylink
            return pylink
        except ImportError:
            if emit_error:
                self.error_occurred.emit("未找到 pylink 库，请安装: pip install pylink")
            return None

    def get_available_devices(self):
        pylink = self._import_pylink(emit_error=False)
        if pylink is None:
            print("[RTT] pylink 导入失败，无法扫描 J-Link 设备")
            return []

        devices = []
        try:
            jlink_temp = pylink.JLink()
            try:
                connected_emulators = jlink_temp.connected_emulators()
                print(f"[RTT] 扫描到 {len(connected_emulators)} 个 J-Link 设备")
            except Exception as e:
                print(f"[RTT] connected_emulators 调用失败: {e}")
                connected_emulators = []

            for emu in connected_emulators:
                try:
                    desc = str(emu)
                    sn = None
                    match = re.search(r'Serial No\.\s*(\d+)', desc)
                    if match:
                        sn = int(match.group(1))

                    if sn is not None:
                        try:
                            jlink_temp.open(serial_no=sn)
                            jlink_name = jlink_temp.product_name if hasattr(jlink_temp, 'product_name') else 'J-Link'
                            jlink_temp.close()
                            devices.append((sn, f"{jlink_name} (SN={sn})"))
                            print(f"[RTT] 发现 J-Link: SN={sn}, 名称={jlink_name}")
                        except Exception as e:
                            print(f"[RTT] 打开 J-Link SN={sn} 失败: {e}")
                            try:
                                jlink_temp.close()
                            except Exception:
                                pass
                            devices.append((sn, desc))
                    else:
                        print(f"[RTT] 无法解析设备序列号: {desc}")
                except Exception as e:
                    print(f"[RTT] 处理设备信息失败: {e}")
                    continue

            if not devices:
                try:
                    jlink_temp.open()
                    sn = jlink_temp.serial_number
                    jlink_name = jlink_temp.product_name if hasattr(jlink_temp, 'product_name') else 'J-Link'
                    jlink_temp.close()
                    devices.append((sn, f"{jlink_name} (SN={sn})"))
                    print(f"[RTT] 通过默认方式发现 J-Link: SN={sn}")
                except Exception as e:
                    print(f"[RTT] 默认方式打开 J-Link 失败: {e}")
                    try:
                        jlink_temp.close()
                    except Exception:
                        pass

        except Exception as e:
            print(f"[RTT] J-Link 扫描异常: {e}")

        print(f"[RTT] 共发现 {len(devices)} 个 J-Link 设备")
        return devices

    def open_connection(self, serial_no=None, chip=None, speed=None, reset_flag=None,
                        start_address=None, range_size=None):
        return super().open_connection(
            serial_no=serial_no,
            chip=chip,
            speed=speed,
            reset_flag=reset_flag,
            start_address=start_address,
            range_size=range_size,
        )

    def _connect_impl(self, serial_no=None, chip=None, speed=None, reset_flag=None,
                      start_address=None, range_size=None):
        pylink = self._import_pylink(emit_error=False)
        if pylink is None:
            raise RuntimeError("未找到 pylink 库，请安装: pip install pylink")

        chip = chip or self.settings.get('chip', 'nRF52840_xxAA')
        speed = speed or self.settings.get('speed', 4000)
        reset_flag = reset_flag if reset_flag is not None else self.settings.get('reset', True)

        if start_address is None:
            addr_str = self.settings.get('start_address', '')
            if addr_str and addr_str.strip():
                start_address = int(addr_str, 16)
        if range_size is None:
            range_str = self.settings.get('range_size', '')
            if range_str and range_str.strip():
                range_size = int(range_str, 16)

        self.jlink = pylink.JLink()
        if serial_no:
            self.jlink.open(serial_no=int(serial_no))
        else:
            self.jlink.open()

        self.jlink.set_tif(pylink.enums.JLinkInterfaces.SWD)
        self.jlink.set_speed(speed)
        self.jlink.connect(chip)

        if reset_flag:
            self.jlink.reset(ms=10, halt=False)

        self.jlink.rtt_start(start_address)

        frame_timeout = self.settings.get('frame_timeout', 50) / 1000.0
        thread = RttReaderThread(
            self.jlink, buffer_idx=0, read_size=8192,
            read_interval=0.002, frame_timeout=frame_timeout,
        )
        self._connect_reader_thread(thread)

    def _close_resource(self):
        if self.jlink and self.jlink.opened():
            try:
                self.jlink.rtt_stop()
            except Exception:
                pass
            self.jlink.close()
        self.jlink = None

    def _send_bytes(self, data: bytes) -> bool:
        offset = 0
        deadline = time.monotonic() + RTT_WRITE_TIMEOUT_SECONDS
        while offset < len(data):
            written = self.jlink.rtt_write(0, list(data[offset:]))
            if not isinstance(written, int) or written < 0 or written > len(data) - offset:
                raise RuntimeError(f"RTT 返回了无效的写入长度: {written}")
            if written == 0:
                if time.monotonic() >= deadline:
                    raise TimeoutError("RTT 写入超时，目标缓冲区可能已满")
                time.sleep(RTT_WRITE_RETRY_INTERVAL_SECONDS)
                continue
            offset += written
        return True

    def get_serial_number(self):
        if self.jlink and self.jlink.opened():
            return self.jlink.serial_number
        return 0

File: src/io/rtt_manager.cpp
#include "rtt_manager.h"
#include "rtt_reader.h"

#include <QDir>
#include <QCoreApplication>
#include <QDebug>

// J-Link constants
#define JLINK_TIF_SWD 1

RttManager::RttManager(QObject *parent)
    : IOTransport(parent)
{
    loadJLinkSDK();
}

RttManager::~RttManager()
{
    closeResource();
    if (m_jlinkLib) {
        m_jlinkLib->unload();
        delete m_jlinkLib;
    }
}

bool RttManager::loadJLinkSDK()
{
    if (m_jlinkAvailable) return true;

    // Try to find JLinkARM.dll in common locations
    QStringList searchPaths;

    // Application directory
    searchPaths << QCoreApplication::applicationDirPath();

    // J-Link installation directory (common locations)
    searchPaths << "C:/Program Files/SEGGER/JLink"
                << "C:/Program Files (x86)/SEGGER/JLink";

    // Environment variable
    QString jlinkDir = qgetenv("JLINK_DIR");
    if (!jlinkDir.isEmpty()) {
        searchPaths << jlinkDir;
    }

    // PATH environment
    QString pathEnv = qgetenv("PATH");
    for (const QString &path : pathEnv.split(';')) {
        if (!path.isEmpty()) {
            searchPaths << path;
        }
    }

    // Try to load DLL (64-bit for our 64-bit application)
    for (const QString &dir : searchPaths) {
        QString dllPath = QDir(dir).absoluteFilePath("JLink_x64.dll");
        if (!QFile::exists(dllPath)) {
            // Fallback to JLinkARM.dll (32-bit)
            dllPath = QDir(dir).absoluteFilePath("JLinkARM.dll");
        }
        if (QFile::exists(dllPath)) {
            m_jlinkLib = new QLibrary(dllPath, this);

            // Load function pointers (JLINKARM_ prefix for most functions)
            fpOpen = (JLINKARM_Open)m_jlinkLib->resolve("JLINKARM_Open");
            fpClose = (JLINKARM_Close)m_jlinkLib->resolve("JLINKARM_Close");
            fpIsOpen = (JLINKARM_IsOpen)m_jlinkLib->resolve("JLINKARM_IsOpen");
            fpGetSN = (JLINKARM_GetSN)m_jlinkLib->resolve("JLINKARM_GetSN");
            fpGetProductName = (JLINKARM_EMU_GetProductName)m_jlinkLib->resolve("JLINKARM_EMU_GetProductName");
            fpGetNumEmulators = (JLINKARM_EMU_GetNumDevices)m_jlinkLib->resolve("JLINKARM_EMU_GetNumDevices");
            fpGetEmuList = (JLINKARM_EMU_GetList)m_jlinkLib->resolve("JLINKARM_EMU_GetList");
            fpTIF_Select = (JLINKARM_TIF_Select)m_jlinkLib->resolve("JLINKARM_TIF_Select");
            fpSetSpeed = (JLINKARM_SetSpeed)m_jlinkLib->resolve("JLINKARM_SetSpeed");
            fpConnect = (JLINKARM_Connect)m_jlinkLib->resolve("JLINKARM_Connect");
            fpSelDevice = (JLINKARM_SelDevice)m_jlinkLib->resolve("JLINKARM_SelDevice");
            fpReset = (JLINKARM_Reset)m_jlinkLib->resolve("JLINKARM_Reset");
            fpHalt = (JLINKARM_Halt)m_jlinkLib->resolve("JLINKARM_Halt");
            fpGo = (JLINKARM_Go)m_jlinkLib->resolve("JLINKARM_Go");
            fpReadMem = (JLINKARM_ReadMem)m_jlinkLib->resolve("JLINKARM_ReadMem");
            fpWriteMem = (JLINKARM_WriteMem)m_jlinkLib->resolve("JLINKARM_WriteMem");
            fpGetDLLVersion = (JLINKARM_GetDLLVersion)m_jlinkLib->resolve("JLINKARM_GetDLLVersion");
            fpGetEmuCaps = (JLINKARM_GetEmuCaps)m_jlinkLib->resolve("JLINKARM_GetEmuCaps");
            fpGetHardwareVersion = (JLINKARM_GetHardwareVersion)m_jlinkLib->resolve("JLINKARM_GetHardwareVersion");
            fpGetFirmwareString = (JLINKARM_GetFirmwareString)m_jlinkLib->resolve("JLINKARM_GetFirmwareString");

            // RTT functions (JLINK_ prefix)
            fpRTT_Control = (JLINK_RTTERMINAL_Control)m_jlinkLib->resolve("JLINK_RTTERMINAL_Control");
            fpRTT_Read = (JLINK_RTTERMINAL_Read)m_jlinkLib->resolve("JLINK_RTTERMINAL_Read");
            fpRTT_Write = (JLINK_RTTERMINAL_Write)m_jlinkLib->resolve("JLINK_RTTERMINAL_Write");

            // Check if essential functions are loaded
            if (fpOpen && fpClose && fpIsOpen && fpConnect &&
                fpRTT_Control && fpRTT_Read && fpRTT_Write) {
                m_jlinkAvailable = true;
                qDebug() << "[RTT] J-Link SDK loaded from:" << dllPath;

                // Get DLL version
                if (fpGetDLLVersion) {
                    int version = fpGetDLLVersion();
                    qDebug() << "[RTT] J-Link DLL version:" << version;
                }

                return true;
            } else {
                qDebug() << "[RTT] Failed to load essential J-Link functions from:" << dllPath;
                delete m_jlinkLib;
                m_jlinkLib = nullptr;
            }
        }
    }

    qDebug() << "[RTT] JLinkARM.dll not found in search paths";
    return false;
}

QList<QPair<QString, QString>> RttManager::getAvailableDevices()
{
    QList<QPair<QString, QString>> devices;

    if (!m_jlinkAvailable) {
        if (!loadJLinkSDK()) {
            return devices;
        }
    }

    QMutexLocker locker(&m_mutex);

    // Get number of emulators
    int numEmulators = 0;
    if (fpGetNumEmulators) {
        numEmulators = fpGetNumEmulators();
        qDebug() << "[RTT] Found" << numEmulators << "J-Link emulators";
    }

    // Enumerate devices by opening each by index
    for (int i = 0; i < numEmulators; ++i) {
        int handle = fpOpen(i);
        if (handle >= 0) {
            int sn = fpGetSN ? fpGetSN() : 0;

            char name[256] = {0};
            if (fpGetProductName) {
                fpGetProductName(name, sizeof(name));
            }

            fpClose();

            QString displayName = QString("%1 (SN=%2)").arg(name).arg(sn);
            devices.append({QString::number(sn), displayName});
            qDebug() << "[RTT] Found J-Link: SN=" << sn << "Name=" << name;
        }
    }

    return devices;
}

QString RttManager::getSerialNumber() const
{
    QMutexLocker locker(&m_mutex);
    return QString::number(m_serialNo);
}

bool RttManager::connectImpl(const QVariantMap &params)
{
    if (!m_jlinkAvailable) {
        emit errorOccurred("J-Link SDK not available. JLinkARM.dll not found.");
        return false;
    }

    QMutexLocker locker(&m_mutex);

    // Get parameters
    QString chip = params.value("chip", "nRF52840_xxAA").toString();
    int speed = params.value("speed", 4000).toInt();
    bool reset = params.value("reset", false).toBool();

    // Open J-Link
    int handle = fpOpen(-1);  // -1 = default device
    if (handle < 0) {
        emit errorOccurred("Failed to open J-Link device");
        return false;
    }

    m_serialNo = fpGetSN ? fpGetSN() : 0;

    // Select SWD interface
    if (fpTIF_Select) {
        int result = fpTIF_Select(JLINK_TIF_SWD);
        if (result < 0) {
            emit errorOccurred("Failed to select SWD interface");
            fpClose();
            return false;
        }
    }

    // Set speed
    if (fpSetSpeed) {
        fpSetSpeed(speed);
    }

    // Select device (chip name)
    if (fpSelDevice && !chip.isEmpty()) {
        QByteArray chipUtf8 = chip.toUtf8();
        int result = fpSelDevice(chipUtf8.constData());
        qDebug() << "[RTT] SelDevice(" << chip << ") result:" << result;
    }

    // Connect to chip
    if (fpConnect) {
        int result = fpConnect();
        if (result < 0) {
            emit errorOccurred("Failed to connect to chip: " + chip);
            fpClose();
            return false;
        }
    }

    // Reset if requested
    if (reset && fpReset) {
        fpReset(10, 0); // ms=10, halt=0
    }

    // Start RTT
    if (fpRTT_Control) {
        int result = fpRTT_Control(1, nullptr); // 1 = Start
        if (result < 0) {
            emit errorOccurred("Failed to start RTT");
            fpClose();
            return false;
        }
    }

    m_connected = true;

    // Create reader thread
    auto *reader = new RttReaderThread(this, 0, 8192, 2, 50);
    connect(reader, &RttReaderThread::dataReceived, this, &IOTransport::dataReceived);
    connect(reader, &RttReaderThread::errorOccurred, this, &IOTransport::errorOccurred);
    startReaderThread(reader);

    return true;
}

void RttManager::closeResource()
{
    QMutexLocker locker(&m_mutex);

    if (m_connected && fpRTT_Control) {
        fpRTT_Control(0, nullptr); // 0 = Stop
    }

    if (fpClose) {
        fpClose();
    }

    m_connected = false;
    m_serialNo = 0;
}

bool RttManager::sendBytes(const QByteArray &data)
{
    if (!m_connected || !fpRTT_Write) {
        return false;
    }

    QMutexLocker locker(&m_mutex);

    int written = fpRTT_Write(0, data.constData(), data.size());
    return written == data.size();
}

int RttManager::readRTT(int bufferIdx, void *data, int size)
{
    if (!m_connected || !fpRTT_Read) {
        return 0;
    }

    QMutexLocker locker(&m_mutex);

    return fpRTT_Read(bufferIdx, data, size);
}
